components/net/fetch/methods.rs

```diff
@@ -866,6 +869,405 @@ async fn scheme_fetch(
+    entries: std::fs::ReadDir,
+) -> BTreeMap<OsString, DirectoryItemDescriptor> {
+    let mut items = BTreeMap::new();
+    for entry in entries {
+        if let Ok(entry) = entry {
+            if let Ok(meta) = entry.metadata() {
+                let os_name = entry.file_name();
+                let entry_name = os_name.to_str().map(str::to_string);
+                if meta.is_dir() {
+                    items.insert(
+                        os_name,
+                        DirectoryItemDescriptor {
+                            item_type: DirectoryItemType::SubDirectory,
+                            name: entry_name,
+                            size: None,
+                            last_modified: Some(meta.modified().map(|m| m.into())),
+                        },
+                    );
+                } else if meta.is_file() || meta.is_symlink() {
+                    items.insert(
+                        os_name,
+                        DirectoryItemDescriptor {
+                            item_type: match meta.is_symlink() {
+                                true => DirectoryItemType::Symlink,
+                                false => DirectoryItemType::File,
+                            },
+                            name: entry_name,
+                            size: Some(meta.len()),
+                            last_modified: Some(meta.modified().map(|m| m.into())),
+                        },
+                    );
+                }
+            }
+        }
+    }
+    items
+}
+
+// TODO: WORK OUT WHY pub(crate) DOES NOT WORK HERE - VISIBILITY SHOULD NOT BE FULLY PUBLIC !!!
+// TODO: FIND A WAY TO MAKE THIS VISIBLE ONLY TO THIS FILE AND THE UNIT TESTS !!!
+pub enum DirectoryItemType {
+    SubDirectory,
+    File,
+    Symlink,
+}
+
+// TODO: WORK OUT WHY pub(crate) DOES NOT WORK HERE - VISIBILITY SHOULD NOT BE FULLY PUBLIC !!!
+// TODO: FIND A WAY TO MAKE THIS VISIBLE ONLY TO THIS FILE AND THE UNIT TESTS !!!
+pub struct DirectoryItemDescriptor {
+    pub item_type: DirectoryItemType,
+    pub name: Option<String>,
+    pub size: Option<u64>,
+    pub last_modified: Option<io::Result<OffsetDateTime>>,
+}
+
+// TODO: WORK OUT WHY pub(crate) DOES NOT WORK HERE - VISIBILITY SHOULD NOT BE FULLY PUBLIC !!!
+// TODO: FIND A WAY TO MAKE THIS VISIBLE ONLY TO THIS FILE AND THE UNIT TESTS !!!
+// Returns an HTML5 document describing the content of the given local directory.
+pub fn build_html_directory_listing(summary: DirectorySummary) -> String {
+    let mut page_html = String::with_capacity(1024);
+    page_html.push_str(
+        "<!DOCTYPE html>\
+<html lang=\"en\">\
+<head><title>Directory listing: ",
+    );
+    let directory_label = match summary.path {
+        Ok(p) => p,
+        Err(e) => format!("&lt;{}&gt;", e),
+    };
+    write_html_safe(&directory_label, &mut page_html);
+    page_html.push_str("</title><style>");
+    page_html.push_str(read_string(Resource::DirectoryListingCSS).as_str());
```

Comment:
I'd recommend against this. For a start, I don't believe that tinytemplate can do what we need, because the HTML is not static, it's procedural (the number of rows, and the types of rows will vary depending on which directory is being viewed). Secondly, we have a value-injection templater a bit like this where I work, and it's very brittle, in that it's easy for a non-developer to accidentally break the structures which must exist within the HTML file to act as hooks for the value injection.

I like the idea of a simple external HTML file which could be edited by non-developers, but I reckon it'll add quite a bit more build time and complexity than is desirable. If you strongly disagree, then I'll continue thinking about it.

Reply:
I've modified the unit tests so that they call "write expected result" methods with the actual data pieces (month, year, etc) instead of specifying the raw, full, expected HTML outcome.

Not as elegant as a single-file template, but it does at least mean that tweaks to the library code HTML generator will now only need a tweak in the relevant "write expected result" method, instead of needing changes to every unit test.